### PR TITLE
feat(service-offering): Check limitations and disable unavailable service offerings (closes #860)

### LINKED ISSUE #860: Manage SOs resource availability in VM creation and SO change dialog
**Acceptance criteria:**

The following rules should be applied to VM creation and to SO change dialog (VM sidebar)
1. **Fixed Service offerings**
- Show Service offerings according to User's resources- disable those offerings that are not available for user
- Add a tooltip for disabled SOs "Not enough CPU/RAM"
2. **Custom Service offerings**
- In Custom SO dialog do not let user to change characteristics for more that are available (change numbers in inputs for maximum ones if user tries to put a bigger number)
3. Check sum of all - current offering + new offering < limit (CPU cores, RAM)

**Connected feature ID:** _vm_creation_so_, _vm_SO_change_

### PR DESCRIPTION
**Pull request by qqcky**

### CODE REVIEW

File: src/app/reducers/vm/redux/vm-creation.effects.ts [hunk 1]
```diff
@@ -305,11 +349,13 @@ export class VirtualMachineCreationEffects {
       this.handleDeploymentMessages(action.payload);
     });
 
-  @Effect({ dispatch: false })
+  @Effect()
   deploymentSuccess$ = this.actions$
     .ofType(vmActions.VM_DEPLOYMENT_REQUEST_SUCCESS)
-    .do((action: vmActions.DeploymentRequestSuccess) => {
+    .map((action: vmActions.DeploymentRequestSuccess) => {
```
Comment (zolotyx) [inline src/app/reducers/vm/redux/vm-creation.effects.ts]: let's leave do and add map also

File: src/app/reducers/vm/redux/vm-creation.effects.ts [hunk 2]
```diff
@@ -118,16 +128,20 @@ export class VirtualMachineCreationEffects {
           const enoughResources = !insufficientResources.length;
 
           return Observable.of(
-            new vmActions.VmCreationEnoughResourceUpdateState(enoughResources),
-            new vmActions.VmInitialZoneSelect(),
+            new vmActions.VmCreationEnoughResourceUpdateState({
+              enoughResources,
+              insufficientResources
+            }),
+            <Action>new vmActions.VmInitialZoneSelect(),
```
Comment (zolotyx) [inline src/app/reducers/vm/redux/vm-creation.effects.ts]: is it necessary?

File: src/app/service-offering/custom-service-offering/custom-offering-restrictions.ts [hunk 3]
```diff
@@ -15,16 +15,16 @@ export interface ICustomOfferingRestrictions {
 
 export const DefaultCustomServiceOfferingRestrictions: ICustomOfferingRestrictions = {
   cpunumber: {
-    min: 0,
-    max: Number.POSITIVE_INFINITY
+    min: 1,
+    max: 2147483647
```
Comment (zolotyx) [inline src/app/service-offering/custom-service-offering/custom-offering-restrictions.ts]: use constant

File: src/app/service-offering/custom-service-offering/custom-service-offering.component.ts [hunk 4]
```diff
@@ -22,17 +22,32 @@ export class CustomServiceOfferingComponent {
...
+      memory: offering.memory || (defaultParams && defaultParams.memory) || restriction.memory.min
     };
     this.restrictions = restriction;
   }
 
+  public setMax(parameter: string, newValue: number) {
+    const max = this.restrictions[parameter].max;
+    if (newValue > max) {
+      this.offering[parameter] = max + 1;
+      setTimeout(() => this.offering[parameter] = max);
```
Comment (zolotyx) [inline src/app/service-offering/custom-service-offering/custom-service-offering.component.ts]: 😢

File: src/app/service-offering/custom-service-offering/custom-service-offering.component.ts [hunk 5]
```diff
@@ -22,17 +22,32 @@ export class CustomServiceOfferingComponent {
...
+      return;
+    }
+    this.offering[parameter] = newValue;
+  }
+
   public errorMessage(lowerLimit: any, upperLimit: any): Observable<string> {
     const upperBound = Number.isInteger(upperLimit);
```
Comment (zolotyx) [inline src/app/service-offering/custom-service-offering/custom-service-offering.component.ts]: please, refactor this, rename variables and maybe remove **any**

File: src/app/service-offering/service-offering-dialog/service-offering-dialog.component.ts [hunk 6]
```diff
@@ -7,12 +7,16 @@ import {
   OnChanges,
   SimpleChanges
 } from '@angular/core';
+import { debug } from 'util';
```
Comment (zolotyx) [inline src/app/service-offering/service-offering-dialog/service-offering-dialog.component.ts]: is this required?

File: src/app/service-offering/service-offering-list/service-offering-list.component.ts [hunk 7]
```diff
@@ -90,22 +103,93 @@ export class ServiceOfferingListComponent implements OnChanges {
...
+  public serviceOfferingTooltip(serviceOffering: ServiceOffering): string {
+    const resources = this.enoughResources(serviceOffering);
+    const tooltips = {
+      memory: 'SERVICE_OFFERING.TOOLTIPS.NOT_ENOUGH_MEMORY',
+      cpu: 'SERVICE_OFFERING.TOOLTIPS.NOT_ENOUGH_CPU',
+      cpuMemory: 'SERVICE_OFFERING.TOOLTIPS.NOT_ENOUGH_MEMORY_AND_CPU'
+    };
+
+
+    if (!resources.enoughMemory) {
```
Comment (zolotyx) [inline src/app/service-offering/service-offering-list/service-offering-list.component.ts]: very strange conditions. is this logic required?

**Comment by bwsw** [comment]
I close that, Looks like very outdated.

**Pull request closed without merging**